### ISSUE TITLE
minifier removes script-type-module attribute. it should not.

### ISSUE THREAD

**Issue by gabor** [DanielRuf/html-minifier-terser#31]
hi,
test case:

input (`test.html`):
```
<script type="module">console.log('hello')</script>
```
command:
```
./node_modules/.bin/html-minifier-terser --remove-script-type-attributes test.html
```

output:
```
<script>console.log('hello')</script>
```

as you see, the `type="module"` part got removed. it should not be removed, because when a script tag has type-module, it's content will be executed differently. also, the documentation for the flag says:
"
  --remove-script-type-attributes      Remove type="text/javascript" from script tags. Other type attribute values are left intact.
"
which is not correct with the behavior.

**Comment by DanielRuf** [comment]
Hi @gabor,

thanks for reporting this. The option at https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L556 runs the following checks:

https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L557

You can see it at https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L67

It checks `executableMimetypes` at https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L169

And this contains more than just `text/javascript`. So the docs are misleading imo.

https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L158-L164

Same at https://github.com/kangax/html-minifier/blob/v4.0.0/src/htmlminifier.js#L163

Also see https://github.com/DanielRuf/html-minifier-terser/pull/29 should probably fix this.

**Comment by DanielRuf** [comment]
Hi @gabor,

thanks for reporting this. The option at https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L556 runs the following checks:

https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L557

You can see it at https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L67

It checks `executableMimetypes` at https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L169

And this contains more than just `text/javascript`. So the docs are misleading imo.

https://github.com/DanielRuf/html-minifier-terser/blob/v5.0.3/src/htmlminifier.js#L158-L164

Same at https://github.com/kangax/html-minifier/blob/v4.0.0/src/htmlminifier.js#L163

Also see https://github.com/DanielRuf/html-minifier-terser/pull/29

**Comment by gabor** [comment]
yes, `module` should probably not be in that list. as far as i can see it got added with https://github.com/DanielRuf/html-minifier-terser/pull/29 ... i asssume the current logic is that for given type-values it minifies-body and removes the type-attribute. for `type=module` the body can probably be minified, but the type-attribute should not be removed.

**Comment by DanielRuf** [comment]
Well, all type=some-valid-js-type are removed and already were in previous versions. The documentation of the option is just outdated.

In my opinion this is not wrong.

**Comment by DanielRuf** [comment]
Both parts are ~4 years old. I doubt that these are wrong. See https://github.com/kangax/html-minifier/blame/v4.0.0/src/htmlminifier.js#L157-L164

**Comment by DanielRuf** [comment]
See https://github.com/kangax/html-minifier/issues/694#issuecomment-235342390 and especially https://github.com/kangax/html-minifier/issues/694#issuecomment-236160075

**Comment by gabor** [comment]
sorry, maybe there is a misunderstanding. what i am saying is that if there is a script tag that looks like this:
```
<script type="module">....</script>
```

then it should not be minified to
```
<script>....</script>
```

browsers handle the content of the script-tag differently when there is a `type="module"` attribute.

you can test it easily, like this:

create a html file with this content:
```
<script>import test from "./test.js"</script>
```

and open it in google-chrome. open the dev-tools, and you will see the following error-message on the console:
"
Uncaught SyntaxError: Cannot use import statement outside a module
"

if you modify the test file to look like this:
"
<script type="module">import test from "./test.js"</script>
"

then it will work. well, it will not work because there is no `./test1.js` file, so you will get a http404 error, or a cors-error if you open it from the filesystem, but you can see you get a different error message.

so, in summary, the 'type="module"' part should not be removed from script-tags.

**Comment by DanielRuf** [comment]
I understood you correctly. This is not a bug per se. See the linked code parts, issues and comments.

The docs are just outdated as the original maintainer wrote and I agree with him here.

**Comment by DanielRuf** [comment]
> so, in summary, the 'type="module"' part should not be removed from script-tags.

In this case you should not use `--remove-script-type-attributes`.

**Comment by gabor** [comment]
i've checked the code mentioned above, but that's not the code from this module.

this link https://github.com/kangax/html-minifier/blame/v4.0.0/src/htmlminifier.js#L157-L164

contains: 

```
var executableScriptsMimetypes = utils.createMap([
  'text/javascript',
  'text/ecmascript',
  'text/jscript',
  'application/javascript',
  'application/x-javascript',
  'application/ecmascript'
]);
```

if i search for this variable in the codebase of this module, i find this:
https://github.com/DanielRuf/html-minifier-terser/blob/master/src/htmlminifier.js#L157

```
var executableScriptsMimetypes = utils.createMap([
  'text/javascript',
  'text/ecmascript',
  'text/jscript',
  'application/javascript',
  'application/x-javascript',
  'application/ecmascript',
  'module'
]);
```

as you see, the last item in the array, the `'module'` is not there in the kangax-version.

it was added in this commit:
https://github.com/DanielRuf/html-minifier-terser/commit/4b9949c288c9a57ea47ff1bfeac781e9b8b5ea99

you can also verify the change in functionality of this program, if you compare version `5.0.2` and `5.0.3`. 

`5.0.2` will not remove `type="module"`, but `5.0.3` will.

**Comment by DanielRuf** [comment]
> it was added in this commit:
> [4b9949c](https://github.com/DanielRuf/html-minifier-terser/commit/4b9949c288c9a57ea47ff1bfeac781e9b8b5ea99)

Yes, please reread my previous comments in detail =)

new script types are supported through this variable array. This is a fact. And this is used by default to remove any valid script types by html-minifier. The logic did not change. Just a new script type is supported now.

As written at https://github.com/DanielRuf/html-minifier-terser/issues/31#issuecomment-588261173:
See https://github.com/kangax/html-minifier/issues/694#issuecomment-235342390 and especially https://github.com/kangax/html-minifier/issues/694#issuecomment-236160075

This is not a bug - the documentation is just wrong. So `wontfix` at the moment. I see no wrong behavior here happening in html-minifier and html-minifier-terser.

Also the original part was / is 4 years old.
https://github.com/DanielRuf/html-minifier-terser/issues/31#issuecomment-588260377

As this is a fork of html-minifier with a few changes it uses the same logic.

**Comment by DanielRuf** [comment]
> it was added in this commit:
> [4b9949c](https://github.com/DanielRuf/html-minifier-terser/commit/4b9949c288c9a57ea47ff1bfeac781e9b8b5ea99)

Yes, please reread my previous comments in detail =)

New script types are supported through this variable array. This is a fact. And this is used by default to remove any valid script types by html-minifier. The logic did not change. Just a new script type is supported now.

As written at https://github.com/DanielRuf/html-minifier-terser/issues/31#issuecomment-588261173:
See https://github.com/kangax/html-minifier/issues/694#issuecomment-235342390 and especially https://github.com/kangax/html-minifier/issues/694#issuecomment-236160075

This is not a bug - the documentation is just wrong. So `wontfix` at the moment. I see no wrong behavior here happening in html-minifier and html-minifier-terser.

Also the original part was / is 4 years old.
https://github.com/DanielRuf/html-minifier-terser/issues/31#issuecomment-588260377

As this is a fork of html-minifier with a few changes it uses the same logic.

**Comment by gabor** [comment]
i did check the comments you linked. one of them is this:

https://github.com/kangax/html-minifier/issues/694#issuecomment-235342390

"We delete whatever default value is for <script type=...>, so that includes application/javascript" .

i completely agree with it, but note, it is the "default" value, not any value.
if you check the html5 spec,  you eventually reach this place:
https://mimesniff.spec.whatwg.org/#javascript-mime-type

as you can see, it is very similar to the list maintained in this module. but, it does not contain "module".

the reason can be found in the html5 spec too:
https://html.spec.whatwg.org/#script-processing-model

"
...
The script's type for a script element is either "classic" or "module".
...
If ...the script element has neither a type attribute nor a language attribute, then...let the script block's type string for this script element be "text/javascript".
...
- If the script block's type string is a JavaScript MIME type essence match, the script's type is "classic".
- If the script block's type string is an ASCII case-insensitive match for the string "module", the script's type is "module".
- If neither of the above conditions are true, then return. No script is executed.
"

so, if no script-type, then that's the same as "text/javascript". that's why it is safe to remove "text/javascript".
then the browser needs to decide if it's "classic" or "module". "classic" is when the type-attribute's value matches with that long list in https://mimesniff.spec.whatwg.org/#javascript-mime-type .and because empty-type-attribute means "classic" too, all those values from https://mimesniff.spec.whatwg.org/#javascript-mime-type can be removed, and the behavior does not change.

but, if the type-string is "module", the script-tag's type is "module", not "classic", and that triggers then different behavior.

this is the reason why i think "module" should not be treated the same as the other possible values.

or, to look at it from the other end: removing script-type-module will break people's code. why should anyone want to remove script-type-module if it breaks their code?

i understand why the above-mentioned pull-request happened. i assume if your script-type is not in the list, the script's contents are not minified, so adding it to the list will allow them to be minified. that is fine. but, removing the "type=module" will break them.

**Comment by DanielRuf** [comment]
There is definitely that may have changed with the recent html(5) specs regarding that as the code parts were not changed in the last 4 years.

There is currently no distinction between such "classic" and "module" type / mode. And as previously mentioned, it is not just `text/javascript` that is already removed.

And https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-type is linked in the comment.

We may have to refactor that.

**Comment by DanielRuf** [comment]
There is definitely something that may have changed with the recent html(5) specs regarding that as the code parts were not changed in the last 4 years.

There is currently no distinction between such "classic" and "module" type / mode. And as previously mentioned, it is not just `text/javascript` that is already removed.

And https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-type is linked in the comment.

We may have to refactor that.

**Comment by DanielRuf** [comment]
@rarous are you interested to take a look at this and improve it?

**Comment by rarous** [comment]
I workaround it by disabling removal of type attribute in html-min settings. It's good for my use cases.

**Comment by DanielRuf** [comment]
@rarous can you share the command that would be needed for the CLI?

**Comment by DanielRuf** [comment]
@rarous can you share the command or setting file + command that would be needed for the CLI?

**Comment by evilebottnawi** [comment]
@DanielRuf Original issue was solved here https://github.com/DanielRuf/html-minifier-terser/commit/4b9949c288c9a57ea47ff1bfeac781e9b8b5ea99, right?

**Comment by DanielRuf** [comment]
@evilebottnawi https://github.com/DanielRuf/html-minifier-terser/issues/31#issuecomment-588996133 ;-)

**Comment by thw0rted** [comment]
I'm having a little trouble following the evolution of this issue.  Is this still considered a "documentation issue" or is the original problem going to be fixed?

It looks to me like the *spirit* of the original behavior was to remove `type="javascript"` and synonyms, because specifying one of those types resulted in the same behavior as omitting the attribute.  `type="module"` has always modified how the tag is interpreted and removing it is always an error.  The current behavior generates broken output any time an ESM module is included.

If the plan is to "refactor" per https://github.com/DanielRuf/html-minifier-terser/issues/31#issuecomment-589082078 then I can just avoid minifying until it's fixed.  If the package maintains that producing broken output is correct behavior then I will update https://github.com/jantimon/html-webpack-plugin/issues/1434 to request that they stop using `removeScriptTypeAttributes` by default.

**Comment by DanielRuf** [comment]
> We may have to refactor that.

So far this is still the case. We have to extend the checks in the minifier and introduce a new array which is not removed but includes the supported types.

PRs which implement this are very welcome.
You can also use ignore comment, see https://github.com/DanielRuf/html-minifier-terser#ignoring-chunks-of-markup

At the moment the docs are incomplete regarding this.

**Comment by DanielRuf** [comment]
<img width="470" alt="Bildschirmfoto 2020-05-12 um 17 52 42" src="https://user-images.githubusercontent.com/827205/81716209-68e4cc00-9479-11ea-8d5a-5e6a0c0f4b9d.png">

Not sure if this is the best solution: https://github.com/DanielRuf/html-minifier-terser/pull/40

I'm not that motivated today so feel free to give constructive feedback / a helpful review.

Not sure if we have to update the documentation.

**Comment by DanielRuf** [comment]
The changes from https://github.com/DanielRuf/html-minifier-terser/commit/dceec7ddab6c7ff2a74cc1f61e6e356c5131c2bf landed in v5.1.1.
